### ISSUE TITLE
PrimaryMappingAttribute to avoid accidentally using the wrong method for mapping an entity.

### ISSUE THREAD

**Issue by PaulVrugt** [riok/mapperly#1008]
**Is your feature request related to a problem? Please describe.**
See https://github.com/riok/mapperly/discussions/1007 . The current implementation of mapping priority (taking the first matching mapping in the file) is unclear and error prone

**Describe the solution you'd like**
I'd like to have an attribute PrimaryMappingAttribute added to the library, that is required to be used (or at least emits a diagnostic) when more than 1 method matches a mapping.

**Describe alternatives you've considered**
I don't see any alternatives. Having it being optional doesn't make sense because you can break your mapping without knowing a different mapper is used, simply because of the position in the file you placed the method.

**Additional context**
https://github.com/riok/mapperly/discussions/1007

**Comment by latonz** [comment]
This is released on the next channel as version `3.5.0-next.2`. It can be used by applying `[UserMapping(Default = true)]`. Any feedback would be appreciated 😊

**Comment by PaulVrugt** [comment]
Let me start by thanking you for taking the effort to fix this issue.
I haven't used the new version in practice, but if I understand correctly how it has now been implemented, I don't think this is very user friendly.

To be able to warnings/errors about ambiguous mappers, I now need to disable the `AutoUserMappings` property on ALL my mappers, and manually mark all user mappers. This has a few issues for me:

- The issue with ambiguous mapping methods is very rare, so I have to write extra code for each mapper a case that rarely happens
- The solution is still very error prone. Because the issue is rare, developers have no indication that they did something wrong if they do not disable the `AutoUserMappings` property. If they later add user mapper method, there is still now indication that it might be using the wrong mapper method.

I think, what we need, is being able to simply leave `AutoUserMappings` enabled, but have the `RMG060` emitted whenever it encounters an ambiguous user mapping. This can then be resolved by marking a user mapping as default.

Please let me know if I misunderstood something, because this doesn't seem like the most robust solution

**Comment by latonz** [comment]
Thank you a lot for your feedback 😊 
You can actually use the `MapperDefaultsAttribute` on the assembly to set `AutoUserMappings` for all mappers in an assembly. See [docs](https://mapperly.riok.app/docs/configuration/mapper/#default-mapper-configuration).
Unfortunately we cannot set the default value of `AutoUserMappings` to false for now since this would be a pretty big breaking change and break a lot of Mappers probably even without notice at compile time. But we plan to do so with the next breaking change (Mapperly 4.0) https://github.com/riok/mapperly/issues/1131. With 4.0 we also plan to increase the default severity of `RMG060` to error. We can't do this now as this may break a lot of users who rely on the current behaviour.

> The issue with ambiguous mapping methods is very rare, so I have to write extra code for each mapper a case that rarely happens
What do you mean by this? Could you elaborate?

> The solution is still very error prone. Because the issue is rare, developers have no indication that they did something wrong if they do not disable the AutoUserMappings property. If they later add user mapper method, there is still now indication that it might be using the wrong mapper method.
The current implementation reports `RMG060` if multiple user-mappings are found for a given type pair without specifying a default, no matter of the value of `AutoUserMappings`. We probably should increase the default severity of `RMG060` from Info to Warn to bring it the attention it needs. A severity of Warn shouldn't break a lot of users.

**Comment by PaulVrugt** [comment]
Thanks for the details!

I'll give it a go with the assembly wide mapper. I might have read the docs wrong. But if rm060 is reported when multiple matching mapping methods are found regardless of the autousermappings value, this solution will probably fit us perfectly, even without disabling autousermappings. I'll get back to you after I've tested the preview version, but it sounds promising 

With "extra code" I actually meant having to set attributes for all places where we use usermappings., but this is not relevant anymore if it works the way I now think it works

**Comment by PaulVrugt** [comment]
Ok,

I've tested it, and to be honest, I think the way it works now is perfect. :D

Whenever mapperly detects there are multiple mappers that can be used, it emits `RMG060`, and we can explicitly set the default by using the `[UserMapping(Default = true)]` attribute.

So the way it works now, we have no reason to disable the `AutoUserMappings` option. And changing the default to false in the future would also be a breaking change for us, and we will have to re-enable it again for all our mappers. 

I might be missing something, but with the current implementation I don't see a reason to change the default for `AutoUserMappings` in the future. Disabling it by default will reintroduce the issue that someone might forget to set the `[UserMapper]` attribute on a method, and mapperly using the wrong mapping function without any warning about this.

Changing the severity for `RMG060` to error in the future does make sense, because otherwise mapperly would need to make an arbitrary decision on what mapper method to use.

In short: 
 - we are happy with the way it works in the current preview version! It prevents mapperly from using the wrong mapper method without notifying us
 - Could you elaborate on why you would want to change the default of `AutoUserMappings` in the future?

**Comment by latonz** [comment]
The idea behind changing the default of `AutoUserMappings` is to transition to a more explicit API by default.

> Disabling it by default will reintroduce the issue that someone might forget to set the [UserMapper] attribute on a method, and mapperly using the wrong mapping function without any warning about this.
This is also true the other way around. Mapperly may use a method never intended to be a mapping method.

But maybe we need to think more about this one, maybe it is not worth to introduce a breaking change here 🤔 We don't plan to release Mapperly 4.0 anytime soon anyway. 

> but we have about 20 projects that use mapperly
We try to keep major releases and their breaking changes as limited as possible. However, sometimes breaking changes are necessary to keep the project moving forward. We try to announce all planned breaking changes early on here https://github.com/riok/mapperly/issues/1131. Everyone is invited to participate in the discussion about whether a particular breaking change is needed (as you just did 😊👍). As you/your company seems to be a heavy user of Mapperly, have you thought about sponsoring the project? 😊

**Comment by PaulVrugt** [comment]
> This is also true the other way around. Mapperly may use a method never intended to be a mapping method.

Well yeah, but if a method is looks like a mapper, and is in a mapper class, and it matches a mapping signature, it probably is a mapper method ;) Otherwise you shouldn't be putting this method in a a mapper class I guess.

> As you/your company seems to be a heavy user of Mapperly, have you thought about sponsoring the project? 😊

Well I might have been exaggerating a bit with 20, but we have a bunch of projects (which are all part of the same product) that all use mapperly. We are using it more and more for new functionality, but Automapper historically is still responsible for a much larget part of the mappings.
